### PR TITLE
Add indicators persistence support

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

We will use any database engine compatible with SQLAlchemy as a persistence layer for calculated indicators.

## Proposal

We choose to integrate the PostgreSQL database in our development configuration but thanks to SQLAlchemy, Warren's API should be compatible with any supported database engine.

- [x] move core base indicators to a dedicated module
- [x] integrate postgresql service
- [x] add persistence related dependencies (SQLModel, Alembic, psycopg)
- [x] add a postgresql service in the CI to test the API

### CODE REVIEW

File: src/api/core/warren/indicators/base.py [hunk 1]
```diff
@@ -92,3 +92,26 @@ def filter_statements(self) -> None:
...
+            f"{hashlib.sha256(self.get_lrs_query().json().encode()).hexdigest()}"
+        )
+
+    async def get_or_compute(self):
+        """Get cached result (if any) or compute the result.
+
+        Nota bene: if computed the result is stored in the database.
+
+        """
+        return 1 + 1
```
Comment (jmaupetit) [inline src/api/core/warren/indicators/base.py:117]: This a WIP. Don't pay attention to this.

File: src/api/core/warren/migrations/README [hunk 2]
```diff
@@ -0,0 +1 @@
+Generic single-database configuration.
```
Comment (jmaupetit) [inline src/api/core/warren/migrations/README:1]: Can be removed?

File: src/api/core/warren/migrations/env.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        context.configure(connection=connection, target_metadata=target_metadata)
+
+        with context.begin_transaction():
+            context.run_migrations()
+
+
+if context.is_offline_mode():
+    run_migrations_offline()
+else:
+    run_migrations_online()
```
Comment (jmaupetit) [inline src/api/core/warren/migrations/env.py:84]: FYI: this file is part of alambic init boilerplate.

File: src/api/core/warren/migrations/script.py.mako [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+branch_labels: Union[str, Sequence[str], None] = ${repr(branch_labels)}
+depends_on: Union[str, Sequence[str], None] = ${repr(depends_on)}
+
+
+def upgrade() -> None:
+    ${upgrades if upgrades else "pass"}
+
+
+def downgrade() -> None:
+    ${downgrades if downgrades else "pass"}
```
Comment (jmaupetit) [inline src/api/core/warren/migrations/script.py.mako:27]: ditto

File: src/api/core/warren/tests/test_base_indicator.py [hunk 5]
```diff
@@ -118,3 +121,27 @@ def test_add_actor_uid():
...
+        def filter_statements(self):
+            pass
+
+        async def compute(self):
+            pass
+
+    indicator = MyIndicator()
+    lrs_query = '{"query_string": null, "query": {"verb": "https://w3id.org/xapi/video/verbs/played"}}'
+    expected = f"myindicator-{hashlib.sha256(lrs_query.encode()).hexdigest()}"
+    assert indicator.cache_key == expected
```
Comment (jmaupetit) [inline src/api/core/warren/tests/test_base_indicator.py:147]: Also part of the WIP to ignore.

File: Makefile [hunk 6]
```diff
@@ -196,11 +197,21 @@ fixtures: \
...
+	@$(COMPOSE) up -d postgresql
+	@$(COMPOSE_RUN) dockerize -wait tcp://$(DB_HOST):$(DB_PORT) -timeout 60s
+	@echo "Create api service database…"
+	@$(COMPOSE) exec postgresql bash -c 'psql "postgresql://$${POSTGRES_USER}:$${POSTGRES_PASSWORD}@$(DB_HOST):$(DB_PORT)/postgres" -c "create database \"warren-api\";"' || echo "Duly noted, skiping database creation."
+	@echo "Running migrations for api service…"
+	@bin/alembic upgrade head
+.PHONY: migrate-api
+
+migrate-app:  ## run django database migrations for the app service
+	@echo "Running api service database engine…"
```
Comment (lebaudantoine) [inline Makefile:211]:
```suggestion
	@echo "Running app service database engine…"
```

File: Makefile [hunk 7]
```diff
@@ -161,7 +162,7 @@ run-app: ## run the app server (development mode)
 run-api: ## run the api server (development mode)
 	@$(COMPOSE) up -d api
 	@echo "Waiting for api to be up and running..."
-	@$(COMPOSE_RUN) dockerize -wait tcp://$(ES_COMPOSE_SERVICE):$(ES_PORT) -timeout 60s
+	@$(COMPOSE_RUN) dockerize -wait tcp://$(DB_HOST):$(DB_PORT) -timeout 60s
```
Comment (lebaudantoine) [inline Makefile:165]: Note: At first read, not sure to get it, discuss it with @jmaupetit.

**Review by lebaudantoine — COMMENTED** [review]
LGTM, I need to test it locally !

**Review by lebaudantoine — APPROVED** [review]

**Comment by lebaudantoine** [comment]
When merging this PR, could be great to close #60 !

**Pull request merged**